### PR TITLE
Aave-v3 DSA position merge

### PR DESCRIPTION
**Pull request by iri031**
Updated function `transferATokens` to merge two DSA Aave positions into single DSA (DSA_A and DSA_B has Aave position, and we need to merge DSA_A into DSA_B).

Contracts updated for: all chains
Script for testing: [Aave-DSA-merge](https://terminal.instadapp.io/q368egrBr)

- [x] Testing ( Polygon, Arbitrum, Avalanche, Optimism)
- [ ] Deployment (Polygon: [0x3827426F69eCf2BAB6F4227dD2C3eC3A2bb6fEC7](https://polygonscan.com/address/0x3827426F69eCf2BAB6F4227dD2C3eC3A2bb6fEC7#code))
- [ ] Enable
- [ ] Update SDK (Polygon: https://github.com/Instadapp/dsa-connect/pull/179)

Tests remaining : 
- Mainnet (waiting for deployment addresses)
- Fantom (flashloan route isn't there)

### CODE REVIEW

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 1]
```diff
@@ -231,22 +238,50 @@ contract AaveHelpers is Helper {
...
+						0,
+						0
+					);
+					if (!getIsColl(_token, address(this))) {
+						aave.setUserUseReserveAsCollateral(_token, true);
+					}
 				}
 			}
+
+			AccountInterface(address(this)).cast(_targets, _data, address(0));
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]:
```suggestion
			AccountInterface(userAccount).cast(_targets, _data, address(0));
```

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 2]
```diff
@@ -199,6 +199,9 @@ contract AaveHelpers is Helper {
 		uint256 rateMode,
 		address user
 	) private {
+		if (instaList.accountID(user) != 0) {
+			user = address(this);
+		}
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]: why?

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 3]
```diff
@@ -275,7 +279,7 @@ contract AaveHelpers is Helper {
 						0,
 						0
 					);
-					if (!getIsColl(_token, address(this))) {
+					if (!getIsColl(_token, userAccount)) {
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]: @Richa-iitr why? We need to enable collateral on `address(this)`

File: contracts/polygon/connectors/aave/v3-import/helpers.sol (hunk 2 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import/helpers.sol]: Made changes

File: contracts/polygon/connectors/aave/v3-import/helpers.sol (hunk 3 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import/helpers.sol]: Got it, updated.

File: contracts/polygon/connectors/aave/v3-import/main.sol [hunk 4]
```diff
@@ -178,7 +187,7 @@ contract AaveV3ImportResolver is AaveHelpers {
 	/**
 	 * @dev Import aave V3 position .
 	 * @notice Import EOA's aave V3 position to DSA's aave v3 position
-	 * @param userAccount The address of the EOA from which aave position will be imported
+	 * @param userAccount The address of the EOA from which aave position will be imported or address of the DSA to which the DSA positions will be merged
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/main.sol]:
```suggestion
	 * @param userAccount The address of the EOA from which Aave position will be imported or the address of the DSA to which the DSA's Aave position will be merged
```

File: contracts/polygon/connectors/aave/v3-import/main.sol [hunk 5]
```diff
@@ -16,11 +16,16 @@ contract AaveV3ImportResolver is AaveHelpers {
...
 	{
-		require(
-			AccountInterface(address(this)).isAuth(userAccount),
-			"user-account-not-auth"
-		);
-
+		if (
+			ListInterface(0x839c2D3aDe63DF5b0b8F3E57D5e145057Ab41556).accountID(
+				userAccount
+			) == 0
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/main.sol]: @Richa-iitr we already have a hardcoded constant variable,right?

File: contracts/polygon/connectors/aave/v3-import/main.sol [hunk 6]
```diff
@@ -3,10 +3,10 @@ pragma solidity ^0.7.0;
 pragma experimental ABIEncoderV2;
 /**
  * @title Aave v3 import connector .
- * @dev  Import EOA's aave V3 position to DSA's aave v3 position
+ * @dev  Import EOA's / withdraw DSA's aave V3 position to DSA's aave v3 position
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/main.sol]:
```suggestion
 * @dev  Import EOA's / External DSA's aave V3 position to DSA's aave v3 position
```

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 7]
```diff
@@ -231,22 +238,50 @@ contract AaveHelpers is Helper {
...
+			string[] memory _targets = new string[](_length);
+			bytes[] memory _data = new bytes[](_length);
+			bytes4 basicWithdraw = bytes4(
+				keccak256("withdraw(address,uint256,address,uint256,uint256)")
+			);
 
-				if (!getIsColl(tokens[i], address(this))) {
-					aave.setUserUseReserveAsCollateral(tokens[i], true);
+			for (uint256 i = 0; i < _length; i++) {
+			_targets[i] = "BASIC-A";
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]:
```suggestion
				_targets[i] = "BASIC-A";
```

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 8]
```diff
@@ -231,22 +238,50 @@ contract AaveHelpers is Helper {
...
+				if (amts[i] > 0) {
+					uint256 _amt = amts[i];
+					address _token = tokens[i];
+					_data[i] = abi.encodeWithSelector(
+						basicWithdraw,
+						_token,
+						_amt,
+						userAccount,
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]:
```suggestion
						address(this),
```

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 9]
```diff
@@ -231,22 +238,50 @@ contract AaveHelpers is Helper {
...
+						0,
+						0
+					);
+					if (!getIsColl(_token, address(this))) {
+						aave.setUserUseReserveAsCollateral(_token, true);
+					}
 				}
 			}
+
+			AccountInterface(userAccount).cast(_targets, _data, address(0));
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]:
```suggestion
			AccountInterface(userAccount).cast(_targets, _data, address(this));
```

File: contracts/polygon/connectors/aave/v3-import/helpers.sol [hunk 10]
```diff
@@ -231,22 +238,50 @@ contract AaveHelpers is Helper {
...
+						"allowance?"
+					);
+					if (!getIsColl(tokens[i], address(this))) {
+						aave.setUserUseReserveAsCollateral(tokens[i], true);
+					}
+				}
+			}
+		} else {
+			string[] memory _targets = new string[](_length);
+			bytes[] memory _data = new bytes[](_length);
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import/helpers.sol]: @Richa-iitr I think first we need to check how many amts are greater than 0 and then assign the length.

File: contracts/polygon/connectors/aave/v3-import/helpers.sol (hunk 9 above)
Comment (thrilok209) [reply contracts/polygon/connectors/aave/v3-import/helpers.sol]: @Richa-iitr Need to add same logics for `_TransferAtokensWithCollateral`

File: contracts/polygon/connectors/aave/v3-import/main.sol (hunk 5 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import/main.sol]: Yes, updating.

File: contracts/polygon/connectors/aave/v3-import/helpers.sol (hunk 9 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import/helpers.sol]: Added.

File: contracts/polygon/connectors/aave/v3-import/helpers.sol (hunk 10 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import/helpers.sol]: Done

**Comment by iri031** [comment]
## Tests [ Polygon ]
Screenshot from 2022-06-02 11-08-56

**Comment by iri031** [comment]
## Tests [ Arbitrum ]
Screenshot from 2022-06-02 14-41-38

**Comment by iri031** [comment]
## Tests [ Avalanche ]

Screenshot from 2022-06-02 14-58-32

## Tests [ Optimism ]

Screenshot from 2022-06-22 17-23-25

**Review by thrilok209 — DISMISSED** [review]
@Richa-iitr Let's deploy on polygon and test it?

**Comment by iri031** [comment]
## Deployed connector for Polygon 

**Name** `AAVE-V3-IMPORT-A`
**Addr** [0x3827426F69eCf2BAB6F4227dD2C3eC3A2bb6fEC7](https://polygonscan.com/address/0x3827426F69eCf2BAB6F4227dD2C3eC3A2bb6fEC7#code)
**tx** [here](https://polygonscan.com/tx/0xd7270c0433a8c3137c0ac453cf1abdc590986176ff56e39ed929d99059f364cd)

**Comment by iri031** [comment]
# Script Results:

 **Migrate DSA_A to DSA_B**

**DSA_A** (before migration)
DSA_A_before

**DSA_B** (before migration)
DSA_B_before

**DSA_A** (after migration)
DSA_A_after

**DSA_B** (after migration)
DSA_B_after

 **Migrate DSA_B to DSA_A** 

**DSA_A** (after migration)
Screenshot from 2022-06-14 15-06-10

**DSA_B** (after migration)
Screenshot from 2022-06-14 17-58-54

**Review by pradyuman-verma — APPROVED** [review]

**Pull request closed without merging**